lib/src/wallet_connect/session.dart

```diff
@@ -0,0 +1,646 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:developer';
+import 'dart:io';
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:protobuf/protobuf.dart';
+import 'package:provenance_dart/proto.dart' as proto;
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/authz/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/bank/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/crisis/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/distribution/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/evidence/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/feegrant/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/gov/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/slashing/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/staking/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/tx/v1beta1/service.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmos/vesting/v1beta1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/cosmwasm/wasm/v1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/google/protobuf/any.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/provenance/attribute/v1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/provenance/marker/v1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/provenance/metadata/v1/tx.pb.dart';
+import 'package:provenance_dart/src/proto/proto_gen/provenance/name/v1/tx.pb.dart';
+import 'package:provenance_dart/src/wallet/crypto/hash/hash.dart';
+import 'package:provenance_dart/src/wallet/encoding/encoding.dart';
+import 'package:provenance_dart/src/wallet/private_key.dart';
+import 'package:provenance_dart/src/wallet_connect/client_meta.dart';
+import 'package:provenance_dart/src/wallet_connect/encrypted_payload_helper.dart';
+import 'package:provenance_dart/src/wallet_connect/messages.dart';
+import 'package:provenance_dart/src/wallet_connect/wallet_connect_address.dart';
+import 'package:uuid/uuid.dart';
+
+extension _DateTimeSecondsSinceEpoch on DateTime {
+  int get secondsSinceEpoch => millisecondsSinceEpoch ~/ 1000;
+}
+
+enum WalletConnectState {
+  connecting,
+  connected,
+  disconnected
+}
+
+class SessionApprovalData {
+  final PrivateKey privateKey;
+  final String chainId;
+
+  SessionApprovalData(this.privateKey, this.chainId);
+}
+
+abstract class WalletConnectionDelegate {
+  Future<List<int>?> onApproveSign(String description, String address, List<int> msg);
```

Comment:
The `on` prefix suggests to me that this is a notification of an approval that has already taken place.